lib/source/pl/core/preprocessor.cpp

```diff
@@ -38,90 +48,83 @@ namespace pl::core {
-        const std::string_view view = &m_code[m_offset];
-        auto end = view.find_first_of(" \t\n\r");
+    std::string getOptionalValue(std::optional<Token> token) {
+        if (!token.has_value())
+            return "";
+        return getTokenValue(token.value());
+    }
 
-        if(end == std::string_view::npos)
-            end = view.length();
+    static bool isNameValid(const std::optional<pl::core::Token> &token) {
+        if (!token.has_value())
+            return false;
+        auto name = getOptionalValue(token);
+        if (name.empty())
+            return false;
+        // replace with std::ranges::all_of when available
+        std::ranges::all_of(name, [](char c) {
+            return std::isalnum(c) || c == '_';
+        });
 
-        m_offset += end;
+        return true;
+    }
 
-        return std::string(view.substr(0, end));
+    bool operator==(const std::vector<pl::core::Token>& a, const std::vector<pl::core::Token>& b) {
+        if (a.size() != b.size())
+            return false;
+        for (u32 i = 0; i < a.size(); i++)
+            if (getTokenValue(a[i]) != getTokenValue(b[i]))
+                return false;
+        return true;
     }
 
-    void Preprocessor::parseComment() {
-        char type = peek(1);
-        if(type == '/') {
-            while(peek() != 0 && peek() != '\n') m_offset++;
-        } else if(type == '*') {
-            char next = peek(2);
-            if(next == '*' || next == '!')
-                return;
-            while(peek() != 0 && (peek() != '*' || peek(1) != '/')) {
-                // update lines
-                if(peek() == '\n') {
-                    m_lineNumber++;
-                    m_lineBeginOffset = m_offset;
-                }
+    void removeValue(std::vector<pl::core::Token> &vec,const std::string &name) {
+        for (unsigned i = 0; i < vec.size(); i++)
+            if (getTokenValue(vec[i]) == name)
+                vec.erase(vec.begin() + i);
+    }
 
-                m_offset++;
-            }
+    std::optional<pl::core::Token> Preprocessor::getDirectiveValue(unsigned line) {
+        const pl::core::Token &token = *m_token;
+        if (m_token->location.line != line || m_token >= m_result.unwrap().end())
+            return std::nullopt;
+        m_token++;
+        return token;
+    }
 
-            m_offset += 2;
-        }
+    std::string Preprocessor::parseDirectiveName() {
+        auto result = m_token->getFormattedValue();
+        m_token++;
+        return result;
     }
 
     void Preprocessor::processIfDef(const bool add) {
         // find the next #endif
         const Location start = location();
         u32 depth = 1;
-        while(peek() != 0 && depth > 0) {
-            if(m_code.substr(m_offset, 6) == "#endif" && !m_inString && m_startOfLine) {
-                m_offset += 6;
+        while(peek() != "EOF" && depth > 0) {
+            if(m_token->getFormattedValue() == "#endif") {
```

Comment:
Instead of comparing the formatted value here, would it be possible to compare the token itself like it's done in other functions?

Reply:
yes, it is more efficient that way. In fact strings shouldn't even be used at this point since the token enums can do a much better job. For example the directive handler map can be from Token::Directive to function and the same for the rest.

Reply:
I removed all the string comparisons and switched to tokens instead even for the directive handler map.